### PR TITLE
Add model MegatronBert

### PR DESCRIPTION
**Pull request by Beacontownfc**
Description
Add new model **MegatronBert**
The model weight:
链接：https://pan.baidu.com/s/1DNoxmqxtRiMycHfVnvwJwg 
提取码：olie

### CODE REVIEW

File: examples/language_model/megatronbert/README.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    --num_train_epochs=2
+```
+训练结束后模型会对模型进行评估，其评估在测试集上完成, 训练完成后你将看到如下结果:
+```text
+eval loss: 0.186327, acc: 0.8992358634742741, eval loss: 0.332409, acc: 0.8968673718470301, eval done total : 118.65499472618103 s
+```
+
+# Reference
+
+* Megatron-LM: Training Multi-Billion Parameter Language Models Using Model Parallelism
```
Comment (Steffy-zxf) [inline examples/language_model/megatronbert/README.md]: 建议同时给出论文链接地址

File: paddlenlp/transformers/megatronbert/modeling.py [hunk 2]
```diff
@@ -0,0 +1,1330 @@
...
+
+    def __init__(self,
+                 vocab_size=29056,
+                 hidden_size=1024,
+                 pad_token_id=0,
+                 type_vocab_size=2,
+                 max_position_embeddings=512,
+                 hidden_dropout_prob=0.1,
+                 position_embedding_type="absolute"):
+        super().__init__()
```
Comment (Steffy-zxf) [inline paddlenlp/transformers/megatronbert/modeling.py]: 为了保持与paddlenlp代码风格统一，建议写成：
```python
super(MegatronBertEmbeddings, self).__init__()
```

File: examples/language_model/megatronbert/args.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        type=str,
+        help="Type of pre-trained model.")
+    parser.add_argument(
+        "--model_name_or_path",
+        default="megatronbert-cased",
+        type=str,
+        help="Path to pre-trained model or shortcut name of model.")
+    parser.add_argument(
+        "--output_dir",
+        default="/root/paddlejob/workspace/output",
```
Comment (Steffy-zxf) [inline examples/language_model/megatronbert/args.py]:
```python
default=None,
```

File: examples/language_model/megatronbert/README.md (hunk 1 above)
Comment (Beacontownfc) [reply examples/language_model/megatronbert/README.md]: Done

File: paddlenlp/transformers/megatronbert/modeling.py (hunk 2 above)
Comment (Beacontownfc) [reply paddlenlp/transformers/megatronbert/modeling.py]: Done

File: examples/language_model/megatronbert/args.py (hunk 3 above)
Comment (Beacontownfc) [reply examples/language_model/megatronbert/args.py]: Done

File: examples/language_model/megatronbert/args.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
+import argparse
```
Comment (ZHUI) [inline examples/language_model/megatronbert/args.py:16]: + copyright

File: examples/language_model/megatronbert/run_squad.py [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+import time
+import json
+import math
+
+from functools import partial
+import numpy as np
+import paddle
+
+from paddle.io import DataLoader
+from args import parse_args
```
Comment (ZHUI) [inline examples/language_model/megatronbert/run_squad.py:26]: 按照 pep8 规则 import. https://www.python.org/dev/peps/pep-0008/#imports
```
Imports should be grouped in the following order:

Standard library imports.
Related third party imports.
Local application/library specific imports.

You should put a blank line between each group of imports.
```

File: paddlenlp/transformers/megatronbert/__init__.py [hunk 6]
```diff
@@ -0,0 +1,2 @@
+from .modeling import *
```
Comment (ZHUI) [inline paddlenlp/transformers/megatronbert/__init__.py]: 该文件可留空

File: paddlenlp/transformers/megatronbert/modeling.py [hunk 7]
```diff
@@ -0,0 +1,1329 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import math
+
+import paddle
+from paddle import nn
+from .. import PretrainedModel, register_base_model
+import paddle.nn.functional as F
```
Comment (ZHUI) [inline paddlenlp/transformers/megatronbert/modeling.py:19]: import 顺序

File: paddlenlp/transformers/megatronbert/modeling.py [hunk 8]
```diff
@@ -0,0 +1,1329 @@
...
+
+def linear_act(x):
+    return x
+
+
+def swish(x):
+    return x * F.sigmoid(x)
+
+
+def gelu_new(x):
```
Comment (ZHUI) [inline paddlenlp/transformers/megatronbert/modeling.py:53]: https://www.paddlepaddle.org.cn/documentation/docs/zh/api/paddle/nn/functional/gelu_cn.html#gelu

gelu with approximate=True？

Comment (Beacontownfc) [reply paddlenlp/transformers/megatronbert/modeling.py:53]: 是近似计算，我是参考了fnet和bigbird的写法

File: examples/language_model/megatronbert/args.py (hunk 4 above)
Comment (Beacontownfc) [reply examples/language_model/megatronbert/args.py:16]: Done

File: examples/language_model/megatronbert/run_squad.py (hunk 5 above)
Comment (Beacontownfc) [reply examples/language_model/megatronbert/run_squad.py:26]: Done

File: paddlenlp/transformers/megatronbert/__init__.py (hunk 6 above)
Comment (Beacontownfc) [reply paddlenlp/transformers/megatronbert/__init__.py]: Done

File: paddlenlp/transformers/megatronbert/modeling.py (hunk 7 above)
Comment (Beacontownfc) [reply paddlenlp/transformers/megatronbert/modeling.py:19]: Done

File: paddlenlp/transformers/megatronbert/modeling.py (hunk 8 above)
Comment (ZHUI) [reply paddlenlp/transformers/megatronbert/modeling.py:53]: 可以替换为，上面的 gelu api，性能会高一些

Comment (Beacontownfc) [reply paddlenlp/transformers/megatronbert/modeling.py:53]: 已替换

File: examples/language_model/megatronbert/README.md [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+```shell
+python -m paddle.distributed.launch run_squad.py \
+    --do_train \
+    --do_predict \
+    --batch_size=8 \
+    --model_name_or_path=megatronbert-cased
+    --learning_rate=1e-5 \
+    --output_dir=output/
+    --device=gpu
+    --num_train_epochs=2
```
Comment (ZHUI) [inline examples/language_model/megatronbert/README.md:27]: shell `\`  有问题

File: examples/language_model/megatronbert/README.md [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+python -m paddle.distributed.launch run_squad.py \
+    --do_train \
+    --version_2_with_negative \
+    --do_predict \
+    --batch_size=8 \
+    --model_name_or_path=megatronbert-cased
+    --learning_rate=1e-5 \
+    --output_dir=output/
+    --device=gpu
+    --num_train_epochs=2
```
Comment (ZHUI) [inline examples/language_model/megatronbert/README.md:60]: 同上

File: examples/language_model/megatronbert/README.md [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+  "best_f1_thresh": -0.20494508743286133
+}
+```
+
+#### 2、mnli数据集
+
+```shell
+python -m paddle.distributed.launch run_glue.py \
+    --task_name=mnli \
+    --output_dir=output/
```
Comment (ZHUI) [inline examples/language_model/megatronbert/README.md]: 同上

File: examples/language_model/megatronbert/args.py [hunk 12]
```diff
@@ -0,0 +1,150 @@
+# Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
Comment (ZHUI) [inline examples/language_model/megatronbert/args.py]: 2022

File: examples/language_model/megatronbert/args.py [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+    parser.add_argument(
+        "--save_steps",
+        type=int,
+        default=5000,
+        help="Save checkpoint every X updates steps.")
+    parser.add_argument(
+        "--seed", type=int, default=42, help="random seed for initialization")
+    parser.add_argument(
+        '--device',
+        choices=['cpu', 'gpu'],
```
Comment (ZHUI) [inline examples/language_model/megatronbert/args.py:108]: 这里没有xpu，上面readme xpu的叙述删掉吧

File: examples/language_model/megatronbert/run_glue.py [hunk 14]
```diff
@@ -0,0 +1,413 @@
...
+from paddlenlp.data import Stack, Tuple, Pad, Dict
+from paddlenlp.data.sampler import SamplerHelper
+from paddlenlp.transformers import MegatronBertTokenizer
+from paddlenlp.transformers import LinearDecayWithWarmup
+from paddlenlp.metrics import AccuracyAndF1, Mcc, PearsonAndSpearman
+from paddlenlp.transformers import MegatronBertForSequenceClassification
+
+FORMAT = '%(asctime)s-%(levelname)s: %(message)s'
+logging.basicConfig(level=logging.INFO, format=FORMAT)
+logger = logging.getLogger(__name__)
```
Comment (ZHUI) [inline examples/language_model/megatronbert/run_glue.py]: 使用 paddlenlp.utils.logger

File: examples/language_model/megatronbert/run_glue.py [hunk 15]
```diff
@@ -0,0 +1,413 @@
...
+        type=int,
+        help="If > 0: set total number of training steps to perform. Override num_train_epochs.",
+    )
+    parser.add_argument(
+        "--seed", default=42, type=int, help="random seed for initialization")
+    parser.add_argument(
+        "--device",
+        default="gpu",
+        type=str,
+        choices=["cpu", "gpu", "xpu", "npu"],
```
Comment (ZHUI) [inline examples/language_model/megatronbert/run_glue.py]: "xpu", "npu" 删除吧，没有验证

File: examples/language_model/megatronbert/run_squad.py [hunk 16]
```diff
@@ -0,0 +1,353 @@
+# Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
Comment (ZHUI) [inline examples/language_model/megatronbert/run_squad.py]: 2022，统一check一下

File: paddlenlp/transformers/megatronbert/modeling.py [hunk 17]
```diff
@@ -0,0 +1,1336 @@
...
+        self.dense = nn.Linear(intermediate_size, hidden_size)
+        self.dropout = nn.Dropout(hidden_dropout_prob)
+
+    def forward(self, hidden_states, input_tensor):
+        hidden_states = self.dense(hidden_states)
+        hidden_states = self.dropout(hidden_states)
+        return input_tensor + hidden_states
+
+
+# Based on transformers.models.bert.modeling_bert.BertLayer. Added LayerNorm.
```
Comment (ZHUI) [inline paddlenlp/transformers/megatronbert/modeling.py]: 注释删除？

File: paddlenlp/transformers/megatronbert/modeling.py [hunk 18]
```diff
@@ -0,0 +1,1336 @@
...
+        This model inherits from :class:`~paddlenlp.transformers.model_utils.PretrainedModel`.
+        Refer to the superclass documentation for the generic methods.
+
+        This model is also a Paddle `paddle.nn.Layer <https://www.paddlepaddle.org.cn/documentation
+        /docs/en/api/paddle/fluid/dygraph/layers/Layer_en.html>`__ subclass. Use it as a regular Paddle Layer
+        and refer to the Paddle documentation for all matter related to general usage and behavior.
+
+        Args:
+            vocab_size (int):
+                Vocabulary size of `inputs_ids` in `ConvBertModel`. Also is the vocab size of token embedding matrix.
```
Comment (ZHUI) [inline paddlenlp/transformers/megatronbert/modeling.py]: `ConvBertModel` conv 统一搜索一下吧

File: paddlenlp/transformers/megatronbert/tokenizer.py [hunk 19]
```diff
@@ -0,0 +1,102 @@
+# Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
Comment (ZHUI) [inline paddlenlp/transformers/megatronbert/tokenizer.py]: 2022

**Review by ZHUI — DISMISSED** [review]
LGTM

File: examples/language_model/megatronbert/README.md (hunk 9 above)
Comment (Beacontownfc) [reply examples/language_model/megatronbert/README.md:27]: Done

File: examples/language_model/megatronbert/README.md (hunk 10 above)
Comment (Beacontownfc) [reply examples/language_model/megatronbert/README.md:60]: Done

File: examples/language_model/megatronbert/README.md (hunk 11 above)
Comment (Beacontownfc) [reply examples/language_model/megatronbert/README.md]: Done

File: examples/language_model/megatronbert/args.py (hunk 12 above)
Comment (Beacontownfc) [reply examples/language_model/megatronbert/args.py]: Done

File: examples/language_model/megatronbert/args.py (hunk 13 above)
Comment (Beacontownfc) [reply examples/language_model/megatronbert/args.py:108]: Done

File: examples/language_model/megatronbert/run_glue.py (hunk 15 above)
Comment (Beacontownfc) [reply examples/language_model/megatronbert/run_glue.py]: Done

File: examples/language_model/megatronbert/run_squad.py (hunk 16 above)
Comment (Beacontownfc) [reply examples/language_model/megatronbert/run_squad.py]: Done

File: paddlenlp/transformers/megatronbert/modeling.py (hunk 17 above)
Comment (Beacontownfc) [reply paddlenlp/transformers/megatronbert/modeling.py]: Done

File: paddlenlp/transformers/megatronbert/modeling.py (hunk 18 above)
Comment (Beacontownfc) [reply paddlenlp/transformers/megatronbert/modeling.py]: Done

File: paddlenlp/transformers/megatronbert/tokenizer.py (hunk 19 above)
Comment (Beacontownfc) [reply paddlenlp/transformers/megatronbert/tokenizer.py]: Done

**Review by ZHUI — DISMISSED** [review]
LGTM

File: paddlenlp/transformers/megatronbert/modeling.py [hunk 20]
```diff
@@ -0,0 +1,1334 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import math
+
+import paddle
+from paddle import nn
+import paddle.nn.functional as F
+
+from ...ops import einsum
```
Comment (yingyibiao) [inline paddlenlp/transformers/megatronbert/modeling.py]: 使用 paddle.einsum

Comment (Beacontownfc) [reply paddlenlp/transformers/megatronbert/modeling.py]: Done

**Review by yingyibiao — APPROVED** [review]
